create_layout.py

```diff
@@ -91,45 +91,120 @@
+    else:
+      removed_files.append(key)
+  for key in after_dict:
+    if key not in before_dict:
+      # Looking for new files
+      added_files.append(key)
+
+  # Returning the snapshot of the new file system
+  return (sorted(unchanged_files), sorted(modified_files), sorted(added_files),
+      sorted(removed_files))
+
+def create_material_rules(changes_of_curr_step, links, index):
   """Create generic material rules (3 variants)
 
+  changes_of_curr_step: changes between the current step's materials and products
+      - a return value of the changes_between_snapshots() function when 
+      the current step's materials and products are the values passed to the function
+  links: an ordered list of in-toto link objects
+  index: the index of the current step
+
   * MATCH available materials with products from previous step (links must be an
   ordered list) and
   * ALLOW available materials if it is the first step in the
   list
+  * DELETE removed materials
+
   Returns a list of material rules
-  NOTE: Read header docstring for ideas for more complexity.  """
+  NOTE: Read header docstring for ideas for more complexity.  
+  """
 
   expected_materials = []
+  unchanged_files, modified_files, _, deleted_files = changes_of_curr_step
+  curr_step_materials = unchanged_files + modified_files + deleted_files
 
   if index == 0:
-    for material_name in links[index].materials.keys():
-      expected_materials.append(["ALLOW", material_name])
+    # If current step is the initial step, we don't need to add MATCH rules.
+    # Instead, ALLOW all the existing files and DELETE deleted files.
+    for file in curr_step_materials:
+      expected_materials.append(["ALLOW", file])
+    for file in deleted_files:
+      expected_materials.append(["DELETE", file])
     expected_materials.append(["DISALLOW", "*"])
 
   else:
-    expected_materials = [
-        ["MATCH", "*", "WITH", "PRODUCTS", "FROM", links[index - 1].name]]
+    # MATCH only those that already were in the previous step
+    # Allow the rest by name
+    pre_step_name = links[index-1].name
+    pre_unchanged_files, pre_modified_files, pre_added_files, _ = \
+        changes_between_snapshots(links[index-1].materials, links[index-1].products)
+    pre_step_products = pre_unchanged_files + pre_modified_files + pre_added_files
+
+    # TODO: the time complexity here is O(n^2). Maybe we can use a hashmap to optimize it.
+    for file in curr_step_materials:
+      if file in pre_step_products:
+        expected_materials.append(
+          ["MATCH", file, "WITH", "PRODUCTS", "FROM", pre_step_name])
+      else:
+        expected_materials.append(["ALLOW", file])
+
+    for file in deleted_files:
+      # DELETE deleted files in the current step
+      expected_materials.append(["DELETE", file])
 
   return expected_materials
 
 
-def create_product_rules(links, index):
-  """Create generic product rules (2 variants)
+def create_product_rules(changes_of_curr_step):
```

Comment:
If we follow the suggestion for `create_material_rules`, we can probably mirror the signature here and call the snapshot function internally.